### PR TITLE
Compress decorated requirement json

### PR DESCRIPTION
**Pull request by SamChou19815**
### Summary

A large source of the big size of the json is caused by repetition of names. For example, if there is a requirement that all CS courses are eligible, then the word `CS` will be repeated for n * m times, where n = number of semesters, m =  number of CS courses.

To optimize this, we can store the classes in a tree structure rather than a flattened giant list. You can see the tree structure type definition in `src/requirements/types.ts`'s `EligibleCourses` type. With this structure, a lot of information are deduplicated! It also allows a much more efficient search, since we can avoid an O(n) search for each requirement.

The result of this optimization is amazing.

### Test Plan 

The json size reduced from 1.4M to less than 700K, and everything still works.

### Notes

Do not merge #77 is merged. I will change the base branch once #77 is merged.

### CODE REVIEW

File: requirement-generator/requirement-satisfiable-course-generator.ts [hunk 1]
```diff
@@ -1,5 +1,12 @@
 import { writeFileSync } from 'fs';
-import { Course, CollegeOrMajorRequirement, DecoratedRequirementsJson } from './types';
+import {
+  Course,
+  CollegeOrMajorRequirement,
+  DecoratedRequirementsJson,
+  UniversityRequirements,
+  DecoratedCollegeOrMajorRequirement,
+  EligibleCourses
+} from './types';
```
Comment (hanywang2) [inline requirement-generator/requirement-satisfiable-course-generator.ts]: Amazing stuff with Course types now that we have a more defined data structure

File: requirement-generator/requirement-satisfiable-course-generator.ts [hunk 2]
```diff
@@ -17,25 +24,50 @@ function courseCanSatisfyCollegeOrMajorRequirement(
...
+): EligibleCourses => {
+  const allMap: { [semester: string]: { [subject: string]: string[] } } = {};
   Object.entries(filteredAllCourses).forEach(([semester, courses]) => {
+    const semesterMap: { [subject: string]: string[] } = {};
     courses
       .filter(course => courseCanSatisfyCollegeOrMajorRequirement(course, collegeRequirement))
-      .forEach(it => set.add(`${semester}: ${it.subject} ${it.catalogNbr}`));
+      .forEach(course => {
+        let subjectSet = semesterMap[course.subject];
+        if (subjectSet == null) {
```
Comment (hanywang2) [inline requirement-generator/requirement-satisfiable-course-generator.ts]: would it be null or undefined?

File: requirement-generator/requirement-satisfiable-course-generator.ts [hunk 3]
```diff
@@ -17,25 +24,50 @@ function courseCanSatisfyCollegeOrMajorRequirement(
```
Comment (hanywang2) [inline requirement-generator/requirement-satisfiable-course-generator.ts]: Once the function is better flushed, we would want documentation for maintainability

File: requirement-generator/requirement-satisfiable-course-generator.ts [hunk 4]
```diff
@@ -17,25 +24,50 @@ function courseCanSatisfyCollegeOrMajorRequirement(
 
 const produceSatisfiableCourseIdForCollegeOrMajorRequirement = (
   collegeRequirement: CollegeOrMajorRequirement
-): readonly string[] => {
-  const set = new Set<string>();
+): EligibleCourses => {
+  const allMap: { [semester: string]: { [subject: string]: string[] } } = {};
```
Comment (hanywang2) [inline requirement-generator/requirement-satisfiable-course-generator.ts]: Can this variable be named a little bit better? I wasn't sure what this is or why it is returned at first glance. `allElidgibleCoursesSemestersMap` would be much better

File: requirement-generator/requirement-satisfiable-course-generator.ts (hunk 2 above)
Comment (SamChou19815) [reply requirement-generator/requirement-satisfiable-course-generator.ts]: `bla == null` is equivalent to `bla === null || bla === undefined`

**Pull request merged**